### PR TITLE
[Driver] SR-3175: Include the terminating signal number in the message of -parseable-output

### PR DESCRIPTION
**Pull request by matthewcarroll**
Add the signal number of the terminated task to the output of -parseable-output. The new key is "signal".
https://bugs.swift.org/browse/SR-3175

In this commit, example output of `swiftc -Xfrontend -debug-crash-immediately` is:

{
  "kind": "signalled",
  "name": "compile",
  "pid": 32822,
  "output": "0  swift ...

  "error-message": "Illegal instruction: 4",
  "signal": 4
`<unknown>:0: error: unable to execute command: Illegal instruction: 4`
`<unknown>:0: error: compile command failed due to signal 4 (use -v to see invocation)`
}

### CODE REVIEW

File: include/swift/Basic/TaskQueue.h [hunk 1]
```diff
@@ -95,11 +95,12 @@ class TaskQueue {
   /// available and SeparateErrors was true.  (This may not be available on all
   /// platforms.)
   /// \param Context the context which was passed when the task was added
+  /// \param Signal the terminating signal number
```
Comment (CodaFi) [inline include/swift/Basic/TaskQueue.h]: As @jrose-apple pointed out in the linked SR, this should mention that it is not available on all platforms (or that is is specifically available on Unix-likes that use that particular implementation)

File: include/swift/Basic/TaskQueue.h [hunk 2]
```diff
@@ -95,11 +95,12 @@ class TaskQueue {
...
   ///
   /// \returns a TaskFinishedResponse indicating whether or not execution
   /// should proceed
   typedef std::function<TaskFinishedResponse(ProcessId Pid, StringRef ErrorMsg,
-                                             StringRef Output, StringRef Errors, void *Context)>
+                                             StringRef Output, StringRef Errors, void *Context, int Signal)>
```
Comment (CodaFi) [inline include/swift/Basic/TaskQueue.h]: This should probably be `Optional<int>`, as the [callsite in the generic queue](https://github.com/apple/swift/blob/master/lib/Basic/Default/TaskQueue.inc#L116) needs updating.

**Review by CodaFi — CHANGES_REQUESTED** [review]

**Comment by jrose-apple** [comment]
To add to @CodaFi's review, this does need a test. I suggest changing test/Driver/crash.swift to check for *some* signal—not a particular one, which may differ between platforms—using a [FileCheck regex](http://www.llvm.org/docs/CommandGuide/FileCheck.html#filecheck-expressions) like `{{-[0-9]+}}`.

**Comment by matthewcarroll** [comment]
Jordan and Robert,

I made the Signal parameter Optional in `TaskSignalledCallback`, updated the header doc to say that the signal may not be available on all platforms, added a test to verify that the signal number is emitted, and added documentation for the new "signal" key emmited in the message of the driver output.

**Comment by matthewcarroll** [comment]
see #6787

**Pull request closed without merging**